kythe/docs/schema/indexing-generated-code.txt

```diff
@@ -31,12 +31,241 @@ implementation, as this is often difficult (or uninteresting) to read. This
 document describes how to encode information about generated code to permit
 cross-language links.
 
+To make the discussion easier to understand let's pretend we are working with
+two languages: SourceLang and TargetLang. SourceLang has `.source` file and TargetLang
+has `.target` files. We also have a tool (generator) that can take generate
+`foo.target` file from `foo.source` file. We have following components:
+
+* Source Indexer - Kythe indexer that takes `.source` files and outputs index
+  data.
+* Target Indexer - Kythe indexer that takes `.target` files and outputs index
+  data.
+* Generator - tool that produces `.target` files from `.source` files.
+* Post processor - Kythe tool that takes all index data produced by all
+  indexers, processes it and outputs final Kythe graph that contains data
+  for both SourceLang and TargetLang.
+
+Now we want to teach Kythe how to create cross-references between generated
+`foo.target` file and original `foo.source` file. The main idea is pretty simple:
+Generator has to output extra data containing mapping of elements in `foo.target`
+to the original elements from `foo.source`. Then when Target Indexer is indexing
+`foo.target` it will use that mapping to output *generates* or *imputes* edges.
+These edges connect nodes from `foo.target` with nodes in `foo.source`.
+
+Kythe doesn't require implementors to use one concrete approach for passing
+mapping metadata and outputting *generates* and *imputes* edges. Below we
+describe two different approaches, each has its own pros and cons. But in
+both cases it is assumed that implementors can change Generator and Target
+Indexer. If possible the *generates* approach is preferred as it requires less
+post-processing work.
+
 TIP: You can find an example implementation at
 link:https://github.com/kythe/kythe/tree/master/kythe/examples/proto[GitHub].
 The current sample web UI does not interpret the parts of the schema we will
 use; this is a work in progress.
 
-== Protocol Buffers
+== Java To JavaScript with *imputes* edges
+
+This approach is generic and works for any combination of SourceLang and
+TargetLang. In this example we generate JavaScript files from Java file so
+SourceLang is Java and TargetLang is JavaScript. Given `Color.java`:
+
+[source,java]
+-------------------------------------------------------------------------------
+public enum Color {
+  RED;
+}
+-------------------------------------------------------------------------------
+
+Generator produces `color.js`:
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+-------------------------------------------------------------------------------
+
+=== Changes to Generator
+
+To support cross-references betwen `color.js` and `Color.java` we need to update
+Generator to output the following mapping data for `Color`, `RED` and `GREEN`
```

Comment:
I don't see any `GREEN` elements.

Reply:
Thanks, it was left from original version where I had RED and GREEN colors but then decided that extra element adds too much noise and having only `Color` and `RED` entities is enough. Fixed.